### DISCUSSION TITLE
Option to demand implementations of interfaces/abstract classes only within the assembly that created them.

### DISCUSSION THREAD

**Discussion by Lucasharskamp** [dotnet/csharplang#8128] [Language Ideas]
- **Introduction**

Let us assume the following context: I have an interface or abstract class that specifies a number of methods and properties, which then will need an entity that implements these things in full. For the sake of security, users of the DLL of this library are not allowed to create their own custom implementations, but rather to use one of the existing implementations made available to them by the library itself. 

I don't know what new keyword would fit, so for now let's call it '_local_' (but I think there are better terms for ''only implemented locally''. _localseal_? _localimpl_?)

This means that the application (or other library) using this assembly is allowed to know about the interface and abstract class and is allowed to use them and call the methods and properties as specified by the contract, but aren't allowed to override them.  

- **Why: prevent unwanted implementations**

This prevents a junior programmer or some third party from abusing the API of the library as it was intended and cause unwanted behavior. Or maybe the code belongs to a firm who wants the behaviors of _IFoo_ and its implementation _Foo_ to be contained in a neat shared library, and not end up with a hundred implementations of _IFoo_ floating around the code of their different products. I've seen this happen at previous employers where _IFoo_ was implemented again and again, although this wasn't intended; it was intended that implementation _Foo_ be used, although this information was lost in time.

- **Why: prevent abuse**

To prevent the worst case scenario, this prevents one from creating a library which doesn't do what the original interface's contract demands at all but rather does something else for nefarious means. (e.g. the interface is about calling a Web API, but the implementation also turns your computer in a bitcoin miner, and since mr. programmer is in a hurry he just grabs the first nuget package he sees of an implementation). 

- **Why: prevent bugs caused by interface contract violation in implementation**

It could also cause bugs where a third party implements the interface, and doesn't do exactly what the original programmers of the interface intended, so when the interface is called from within the original assembly, it could cause bugs.

Example:
```
public interface IFoo
{
     public string BarWithValue();
}
```

In this example, _BarWithValue()_ shouldn't return _null_ since it would disrupt the behavior of the original assembly. So let's assume a third party does the following:

```
// somewhere in the original Library where IFoo comes from.
public static class LibraryServiceClass 
{
     public void userDecicedService(IFoo foo) 
     {
           IFoo userDecicedService = foo;
           SetBar(userDecicedService .BarWithValue()) // throws exception if null was provided.
     }
}

...

// application using the library, written by a third party:
public class Foo : IFoo
{
     public string BarWithValue() { return null; }
}
...
// application using this library
LibraryServiceClass.SetFooHandler(new Foo());

```

In this case,  a third party misused the library, and we only want them to enact behaviors that we allow as specified by the interface contract.

- **The solution: allow the library to demand local implementations only to ensure interface contract adherence**

 For interfaces, implementing this is simple:

```
public local interface IFooInterface 
{
   public string Bar();
}
```

Examples:

```
// in the original library:
public class FooImplementation : IFooInterface // this is fine!
{
     public string Bar() { return "X"; }
}
...

// in the assembly that uses this interface:
IFooInterface a = new FooImplementation(); 
return a.Bar(); // this is fine!

```
this would be allowed, since we're using the implementations as specified in the API of the library we're using. _FooImplementation_ was created in the original library as well. The using assembly knows what _IFooInterface_ is, it's not hidden and we are allowed to use its properties and methods as-is. The only thing we're not allowed to do is override them. The following would cause an error:

```
public class NewImplementationForFoo : IFooInterface // Error: cannot implement local interfaces outside the assembly that established them!
```

If the entire _interface_ or _abstract class_ is marked _local_ this is all pretty simple; use them as much as you like since it's simply a contract, but you're not going to make your own implementation.

If you want to mark specific properties or methods as local, however, this can be a bit more specific: for example:

```
public abstract class Foo
{
      public abstract void Bar();
      public local abstract void FooBarOne();
      public local virtual void FooBarTwo() { .. }
}
```

If we were to override the _Foo_ class outside the assembly where it was established, this would mean we cannot override an existing implementation of _Foo_: we can override _Bar_, but not _FooBarOne_

That would make the following illegal if _FooImpl_ isn't part of the assembly as _Foo_:

```
public class FooImpl : Foo // Error: 'Foo' contains methods that are marked as only overridable by the local assembly!
{
      public override void Bar() { ... }
      public override void FooBarOne() { .. }
}
```

This means that, if we want to expose _abstract class FooImpl_ to the outside world, the methods marked _local_ from the original would need to have been implemented. That means the following (where _FooImpl_ is from the same assembly as _Foo_) is legal behavior: 

```
// in the assembly where Foo was established:
public abstract class FooImpl : Foo
{
      public local override void FooBarOne() { .. }
     // FooBarTwo doesn't need to be implemented since that is a virtual function that already contained a body
}

// in the third party application where we use FooImpl:
public sealed class FooFinal: FooImpl  // this is fine!
{
      public override void Bar() { ... } // this is fine!
      public override void FooBarTwo() { .. } // this is not fine!
}

```

**Note:** the overridden methods marked local must be marked as either _local_ again or as _sealed_, for obvious reasons.

- **Potential issues**

1. If the library doesn't create implementations of these local interfaces or abstract classes (or their methods/properties), it renders them unusable. Upon compilation, a warning or error must be given that no implementations exist that one can use when using the library.
2. Compiler Warning/Info: Specifying _local_ methods/properties is kind of redundant if the entire interface/abstract class is specified _local_ as well.
3. Compiler Error/Warning/Info: trying to define a _private_ or _internal_ entity as _local_, or an entity that cannot be overridden anyways as _local_. For example: `public sealed local override FooBarImplementation() { .. }` is kind of redundant.
4. Another question is whether to allow extension methods for local interfaces or abstract classes, although I don't think that should be a problem and should be allowed since they're not part of the original object anyway.

**Comment by HaloFour** [comment]
If you declare an `internal` constructor on an abstract class that will force it so that only classes within the same assembly can derive from that abstract class.

**Reply by Lucasharskamp** [reply]
That would be nice, where it not for the following issues:
1. Dependency injections don't swallow non-public constructors. The whole reason I'm opening this ticket is actually because of dependendy injection services; yes, I've made an implementation that is an _internal class_ and that can only be set up with an extension method which sets up the service, but I don't want third parties to implement the contract anyways.
2. sometimes you'd want an interface rather than an abstract class, for things like _struct_'s or _record_'s, especially when we're doing object composition over simple inheritance (e.g. using a dozen interfaces for a single class) and you want to ensure multiple interfaces aren't implemented outside the library.

**Reply by CyrusNajmabadi** [reply]
The point of an interface is to explicitly state that as long as you abide by the contract of the interface, then things are good. 

If you want to limit to a particular impl, then use an abstract class. 

If you want an interface and you want to disallow other implementations of it, then write an analyzer that errors if a user does that.

**Reply by CyrusNajmabadi** [reply]
> but I don't want third parties to implement the contract anyways.

So have an internal constructor.  Third parties will not be able to implement it.

**Reply by CyrusNajmabadi** [reply]
> sometimes you'd want an interface rather than an abstract class, for things like struct's or record's, especially when we're doing object composition over simple inheritance (e.g. using a dozen interfaces for a single class)

Make the interface internal. Have strong typed entry points for all the supported types. Delegate internally to working on the interface. 

This has the added benefit of the API being clear about the concrete types it takes, and preventing users from thinking they can implement the interface themselves.

**Reply by Lucasharskamp** [reply]
> The point of an interface is to explicitly state that as long as you abide by the contract of the interface, then things are good.

No garantuees there.  I want people to use the interface, not override them for the reasons listed above.

> If you want to limit to a particular impl, then use an abstract class.

The point of an abstract class is that people can define specific parts of it (abstract), or replace them (virtual). The problem is, in C# the acessibility for overrides is the same as who is allowed to use the method. If it was ''override is internal, access to the method's use is public'' we wouldn't be having this conversation.

>So have an internal constructor.

Again, dependency injections. (And there are cases where we should allow initialisation of the object via public parameterless constructors)

The difference is the access of the usage of the object or interface and its methods/properties, and acess to defining the behaviors of it.

**Reply by CyrusNajmabadi** [reply]
> I want people to use the interface, not override them for the reasons listed above.

Then add an analyzer. 

> The problem is, in C# the acessibility for overrides is the same as who is allowed to use the method. If it was ''override is internal, access to the method's use is public'' we wouldn't be having this conversation.

If the constructor is internal, they can't override anything since they can't subclass the type. 

> Again, dependency injections.

I didn't know what DI changes here.

**Reply by Lucasharskamp** [reply]
>Make the interface internal. Have strong typed entry points for all the supported types. Delegate internally to working on the interface.

Again, dependency injection. Let's assume the following: we're using a dependency injection of interface _IFoo_ with implementation class _Foo_. The service itself is handled by extending the service collection and calling said extension method in the program we're we are using the library:
```
public static IServiceCollection SetupFoo(this IServiceCollection services) { return services.AddSingleton<IFoo, Foo>(); }
```
Using your example, I wouldn't be able to call _IFoo_ since _IFoo_ is marked as internal. And since the injection is always the left-hand side object, I can't use it in the application where I'm implementing the dependency injection. And no, we simply cannot let IFoo out the door and just use Foo: there are other DI's which use IFoo's methods, but have a different implementation (e.g. _FooTwo_)

**Reply by Lucasharskamp** [reply]
>I didn't know what DI changes here.

Please read before answering: as mentioned before, you cannot use non-public constructors for Di's. And there are also cases where you'd want a public parameterless constructor.

**Reply by HaloFour** [reply]
Nothing limits the constructors of the implementing classes from being public.  Dependency injection should be fine with this strategy, given the DI container needs to know which implementation goes with the contract anyway.

The compiler and language don't really care about the limitations of particular dependency injection containers.  It'd be on them to correctly handle how the language and runtime implement things.

**Reply by 333fred** [reply]
DI isn't going to be making an instance of the _abstract_ class anyway, so I don't understand the resistance to that solution? You can derive from the abstract class and mark that sealed; that derived type can have a public constructor, no one can inherit it, and no one external can inherit the abstract class because of the internal constructor.

**Reply by CyrusNajmabadi** [reply]
> Please read before answering: 

I did read it. 

> as mentioned before, you cannot use non-public constructors for Di's. And there are also cases where you'd want a public parameterless constructor.


I don't see what that has to do with abstract class. That might be a requirement of your DI system for the derived types.  But those are not affected by what I am saying. 

Furthermore, this appears to an issue with the DI system you're using.  We can't provide solutions at the language level because, for all we know, that DI system simply will not support anything we create.

**Comment by colejohnson66** [comment]
The best you can get (out of the box) would be a custom analyzer that raises errors when library consumers implement your interface.

However, there is another, more interesting way, that can be used: IL rewriting. Avalonia, for example, have an attribute [\[NotClientImplementable\]](https://github.com/AvaloniaUI/Avalonia/blob/5e323b8fb1e2ca36550ca6fe678e487ff936d8bf/src/Avalonia.Base/Metadata/NotClientImplementableAttribute.cs#L14) they tag to interfaces. Then, during build, they [inject in a method with an "unspeakable" name, `(This interface or abstract class is -not- implementable by user code !)`](https://github.com/AvaloniaUI/Avalonia/blob/5e323b8fb1e2ca36550ca6fe678e487ff936d8bf/nukebuild/RefAssemblyGenerator.cs#L76-L90).

If you try to implement any of those interfaces, you get a compiler error about the method you "forgot" to implement:

```csharp
public class MyCustomPen : IPen
{
    public IBrush? Brush { get; }
    public IDashStyle? DashStyle { get; }
    public PenLineCap LineCap { get; }
    public PenLineJoin LineJoin { get; }
    public double MiterLimit { get; }
    public double Thickness { get; }
}
```

Produces:

```
Interface member 'void Avalonia.Media.IPen.(This interface or abstract class is -not- implementable by user code !)()' is not implemented
```
